### PR TITLE
Add selecting tags on the compare page

### LINKED ISSUE #3780: Feature: Allow comparison of tags, etc.
- Gitea version (or commit ref): current
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes, e.g. https://try.gitea.io/hans/testa/src/tag/funny

## Description

Right now the comparison functionality is unnecessarily constraint. It only allows to compare branches, but not tags or single commits. Comparing to tags would be a very useful feature; e.g. to view changes since last release, or compare different releases.

1. Allow to provide tag names for comparison, e.g. https://try.gitea.io/hans/testa/compare/master...funny or https://try.gitea.io/hans/testa/compare/master...tag:funny

2. Provide green comparison button for tags as well as branches, e.g. the button is currently present here:
https://try.gitea.io/hans/testa/src/branch/master
but not here:
https://try.gitea.io/hans/testa/src/tag/funny

3. While we are at it, also provide the functionality for commit hashes, e.g.
https://try.gitea.io/hans/testa/compare/bd0928b003...master

This is also related to #1352

### PR DESCRIPTION
**Pull request by jtran**
Related to #3780, #1352, #12251.

On the `/compare` page, before you could only select branches in the UI in order to make a pull request.  This PR makes it so you can also select tags to compare them.  The comparison already exists; this just provides an easier UI for navigating to those comparison pages.

When branches are selected and you have permission to create a pull request, the display hasn't changed.
Screen Shot 2021-05-04 at 2 59 00 PM

When a tag is selected for either base or head, the display changes to comparing.
Screen Shot 2021-05-04 at 2 59 44 PM

This is the dropdown, with both branches and tags, which now matches the dropdown on other pages.
Screen Shot 2021-05-04 at 3 33 41 PM

### Review

For easier reviewing, I made a separate commit that changes indentation.

close  #3780

### CODE REVIEW

File: integrations/compare_test.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+)
+
+func TestCompareTag(t *testing.T) {
+	defer prepareTestEnv(t)()
+
+	session := loginUser(t, "user2")
+	req := NewRequest(t, "GET", "/user2/repo1/compare/v1.1...master")
+	resp := session.MakeRequest(t, req, http.StatusOK)
+	htmlDoc := NewHTMLParser(t, resp.Body)
+	selection := htmlDoc.doc.Find(".ui.choose.branch .filter.dropdown")
```
Comment (silverwind) [inline integrations/compare_test.go]:
```suggestion
	selection := htmlDoc.doc.Find(".choose.branch .filter.dropdown")
```

File: web_src/js/index.js [hunk 2]
```diff
@@ -1241,10 +1241,16 @@ async function initRepository() {
     $(this).select();
   });
 
+  // Compare or pull request
+  const $repoDiff = $('.repository.diff');
+  if ($repoDiff.length > 0) {
```
Comment (silverwind) [inline web_src/js/index.js]:
```suggestion
  if ($repoDiff.length) {
```

File: web_src/js/index.js [hunk 3]
```diff
@@ -3433,6 +3439,17 @@ function initIssueTimetracking() {
   });
 }
 
+function initBranchOrTagDropdown(selector) {
+  $(selector).each(function() {
+    const $dropdown = $(this);
+    $dropdown.find('.reference.column').on('click', function () {
+      $dropdown.find('.scrolling.reference-list-menu').css('display', 'none');
+      $($(this).data('target')).css('display', 'block');
```
Comment (silverwind) [inline web_src/js/index.js]:
```suggestion
      $dropdown.find('.scrolling.reference-list-menu').hide();
      $($(this).data('target')).show();
```

Please test it.

**Review by silverwind — APPROVED** [review]

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
just noticed one nit, but it isnt introduced by this pull ... loading the dialog to select branches/tags can take quite some time on large repos :/

**Pull request merged**